pkg/l2announcer: ensure leases are only created for services that are being announced

f1ko opened the pull request:
pkg/l2announcer: ensure leases are only created for services that are being announced

Previously leases were created for all services in the cluster regardless of them having an announceable IP or not.

This change ensures that these services are skipped and only services with an external and/or LB IP - depending on the policy - will have a lease. Fixes: https://github.com/cilium/cilium/issues/28752

```release-note
l2announcer: Leases are only created for services that are being announced.
```

dylandreimerink commented on pkg/l2announcer/l2announcer.go:
```diff
@@ -311,8 +311,19 @@ func (l2a *L2Announcer) processPolicyEvent(ctx context.Context, event resource.E
+	noLB := true
+	for _, v := range svc.Status.LoadBalancer.Ingress {
+		noLB = noLB && v.IP == ""
+	}
```
This reads odd. I suggest writing it like so which is a more familiar pattern
```suggestion
	noLB := true
	for _, v := range svc.Status.LoadBalancer.Ingress {
		if v.IP != "" {
		    noLB = false
		    break
		}
	}
```

dylandreimerink replied:
The rest of the changes look good to me!

f1ko replied:
Thanks for pointing it out. Changed in https://github.com/cilium/cilium/pull/29446/commits/26d1c28a38fbb17998aa94b76dce4a3cbf4629dc.

youngnick commented on pkg/l2announcer/l2announcer.go:
```diff
@@ -311,8 +311,19 @@ func (l2a *L2Announcer) processPolicyEvent(ctx context.Context, event resource.E
+	noExternal := svc.Spec.ExternalIPs == nil
+	noLB := true
+	for _, v := range svc.Status.LoadBalancer.Ingress {
+		noLB = noLB && v.IP == ""
```
This took me a bit to parse - I haven't seen this sort of assignment used before. Is this a common Cilium pattern?

f1ko replied:
Thanks for pointing it out. Changed it in https://github.com/cilium/cilium/pull/29446/commits/26d1c28a38fbb17998aa94b76dce4a3cbf4629dc.

youngnick reviewed (approved):
Thanks for that, this LGTM.

f1ko committed:
pkg/l2announcer: only create leases for services that are being announced

Previously leases were created for all services in the cluster regardless of them having an announceable IP or not.

This change ensures that these services are skipped and only services with an external and/or LB IP - depending on the policy - will have a lease. Fixes: https://github.com/cilium/cilium/issues/28752

Signed-off-by: Filip Nikolic <oss.filipn@gmail.com>
